azurelinuxagent/ga/utils/exthandler_utils.py

```diff
@@ -0,0 +1,343 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+
+
+from azurelinuxagent.common import logger as logger, conf
+from azurelinuxagent.common.event import WALAEventOperation, add_event
+from azurelinuxagent.common.exception import ExtensionHandlerConfigurationError
+from azurelinuxagent.common.future import ustr
+from azurelinuxagent.common.osutil import get_osutil
+from azurelinuxagent.common.version import AGENT_NAME
+
+MEMORY_OOM_KILL_OPTIONS = ["enabled", "disabled"]
+DEFAULT_CORES_COUNT = -1
+
+AGENT_CGROUP_NAME = "WALinuxAgent"
+DEFAULT_CPU_LIMIT_AGENT = 10.0
+DEFAULT_CPU_LIMIT_EXT = 40.0
+DEFAULT_MEM_LIMIT_MIN_MB_FOR_EXTN = 256.0  # mb, applies to agent and extensions
+DEFAULT_MEM_LIMIT_MAX_MB_FOR_AGENT = 512.0  # mb, applies to agent only
+DEFAULT_MEM_LIMIT_PCT_FOR_EXTN = 15.0  # percent, applies to extensions
+
+CGROUPS_ENFORCE_DEFAULT_LIMITS = False  # Currently, we don't want to enforce defaults as well.
+
+
+class HandlerConfiguration(object):
+    @staticmethod
+    def send_handler_configuration_event(name=AGENT_NAME, message="", operation=WALAEventOperation.Unknown,
+                                         is_success=True, log_event=True):
+        if message:
+            add_event(name=name, message=message, op=operation, is_success=is_success, log_event=log_event)
+
+    def __init__(self, handler_config_json, name=None):
+        self._data = handler_config_json
+        self._name = name
+        self.resource_config = None
+        self.resource_limits = None
+
+        try:
+            if handler_config_json is None:
+                raise ExtensionHandlerConfigurationError('Malformed handler configuration file. '
+                                                  'No data present')
+            if "handlerConfiguration" not in handler_config_json:
+                raise ExtensionHandlerConfigurationError('Malformed handler configuration file. '
+                                                  '"HandlerConfiguration" key not present')
+            if "linux" not in handler_config_json['handlerConfiguration']:
+                raise ExtensionHandlerConfigurationError('No linux configurations present in '
+                                                  'HandlerConfiguration')
+
+            try:
+                self.resource_config = ExtensionResourcesConfiguration(self.get_linux_configurations())
+            except ExtensionHandlerConfigurationError as e:
+                raise e
+
+            if self.resource_config or CGROUPS_ENFORCE_DEFAULT_LIMITS:
```

Comment:
shouldn't this be 'and' (instead or 'or')

can we add a unit test for this condition?

Reply:
Added. `test_ext_handler_with_cgroup_invalid_limits_passed_defaults_set` and `test_ext_handler_with_no_limits_passed_defaults_set`.